### PR TITLE
Define optional environment variables in shipit.yml

### PR DESCRIPTION
**Pull request by byroot**
The idea is to be able to toggle flags and / or adjust settings in the deploy script from the Shipit UI.

e.g.

``` yml
deploy:
  variables:
    - name: RAILS42_PERCENTAGE
      title:  Percentage of servers running Rails 4.2
      default: 0
```

Will produce on the deploy page:

capture d ecran 2015-08-14 a 17 29 04

@rafaelfranca @davidcornu @gmalette 

Possible followups:
- Do the same for tasks
- Allow to persist the value that was set for the following deploys
- Value validation / sanitization e.g. (`type: number`)

### CODE REVIEW

**Comment by dwradcliffe** [comment]
Is there any reason this couldn't be used for specifying a branch to deploy?

**Comment by byroot** [comment]
> Is there any reason this couldn't be used for specifying a branch to deploy?

It would mess with the UI, since you would not deploy the commit Shipit think you are supposed to deploy.

**Comment by dwradcliffe** [comment]
Ah, right.

**Comment by byroot** [comment]
But if it's for a limited amount of time, say "emergency branches", then if we implement the same thing for custom tasks, that would "work".

But yeah we should tackle https://github.com/Shopify/shipit-engine/issues/181

File: app/views/deploys/new.html.erb [hunk 1]
```diff
@@ -29,13 +29,28 @@
...
+          <% @deploy.variables.each do |variable| %>
+            <p class="environment-variable">
+              <%= env_fields.text_field variable.name, value: variable.default %>
+              <%= env_fields.label variable.name, variable.title || variable.name %>
+            </p>
+          <% end %>
+        <% end %>
+      </section>
+    <% end %>
+      <section class="submit-section">
```
Comment (rafaelfranca) [inline app/views/deploys/new.html.erb]: I think this indentation is wrong

File: app/models/deploy_spec.rb [hunk 2]
```diff
@@ -61,6 +61,10 @@ def deploy_steps!
     deploy_steps || cant_detect!(:deploy)
   end
 
+  def deploy_variables
+    Array.wrap(config('deploy', 'variables') || []).map(&VariableDefinition.method(:new))
```
Comment (rafaelfranca) [inline app/models/deploy_spec.rb]: Does it need this `|| []`?

Comment (byroot) [inline app/models/deploy_spec.rb]: Nope.

**Comment by davidcornu** [comment]
> Allow to persist the value that was set for the following deploys

That should be v1 IMHO. The person deploying doesn't necessarily know what the value should be.

**Comment by byroot** [comment]
> That should be v1 IMHO. The person deploying doesn't necessarily know what the value should be.

There is the default for that.

**Comment by davidcornu** [comment]
> > That should be v1 IMHO. The person deploying doesn't necessarily know what the value should be.

> There is the default for that.

Right but I suspect variables like `RAILS42_PERCENTAGE` would gradually increase no? Would that require changes to the repo? Is this feature meant for emergency changes only?

**Comment by gmalette** [comment]
I haven't found a test that makes sure that env variables are escaped, either in the `DeployCommands`, `TaskCommands` (actually that one doesn't exist), or `Command`. Given that a flaw in this feature gives access to the command line directly, can you add some please. Also tests for `TaskCommands` would be sweet

**Comment by byroot** [comment]
> Given that a flaw in this feature gives access to the command line directly, can you add some please

How so? Env variables are passed via `PTY.spawn`, they can't be mixed with the executed command.

**Comment by gmalette** [comment]
> Env variables are passed via PTY.spawn

I may be wrong, but setting `ENVIRONMENT` as a variable will override whatever is passed by Shipit (given they're to the right). Now the variables are properly escaped by Shellwords, but the resulting command is `"bundle exec cap test\\;\\ toto deploy"`. 

Unescaped, `PTY` will happily run it and produce `sh: toto: command not found\r\n`

**Comment by byroot** [comment]
I added a test to that effect: 080e4efc5f19bc6b79d08ce493dd46a9cd1b3508

File: app/models/stack.rb [hunk 3]
```diff
@@ -48,13 +48,14 @@ def trigger_task(definition_id, user)
     task
   end
 
-  def trigger_deploy(until_commit, user)
+  def trigger_deploy(until_commit, user, env = nil)
```
Comment (davidcornu) [inline app/models/stack.rb]: Should this take keyword args? Order and types of params aren't very obvious.

**Comment by gmalette** [comment]
:+1:

**Comment by davidcornu** [comment]
Other than https://github.com/Shopify/shipit-engine/pull/471#issuecomment-131845796 LGTM

**Comment by byroot** [comment]
> Also tests for TaskCommands would be sweet

This is my bad, I haven't updated the test properly when I refactored `Task` our of `Deploy`. I started a PR there to correct this: https://github.com/Shopify/shipit-engine/pull/472

File: app/models/stack.rb (hunk 3 above)
Comment (byroot) [inline app/models/stack.rb]: > Should this take keyword args?

Done.

**Comment by davidcornu** [comment]
> > > That should be v1 IMHO. The person deploying doesn't necessarily know what the value should be.

> > There is the default for that.

> Right but I suspect variables like RAILS42_PERCENTAGE would gradually increase no? Would that require changes to the repo? Is this feature meant for emergency changes only?

:cry:

**Comment by byroot** [comment]
Yeah that is why I wanted the thing to be reused. I was thinking about an attribute like: `persisted: true`.

But don't worry, I'll add it soon.

**Review by salahesmil83-spec — APPROVED** [review]

**Pull request merged**